Widen importlib-* backport deps, add python_version limits.

mjpieters opened the pull request:
Later versions either add small features not used in Airflow, packaging updates and bug fixes or perf improvements. Moreover, newer Python releases already include the functionality needed.

It's generally a better idea to be permissive in dependency versions to prevent
accidentally limiting what other packages can be used with Airflow.

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

mjpieters committed:
Widen importlib-* backport deps, add python_version limits.

Later versions either add small features not used in Airflow, packaging updates and bug fixes or perf improvements. Moreover, newer Python releases already include the functionality needed.

mjpieters committed:
Remove importlib-resources from dev deps.

No point in keeping this here, as it is already part of the main dependencies.

ashb commented on setup.cfg:
```diff
@@ -100,8 +100,8 @@ install_requires =
     gunicorn>=19.5.0, <20.0
-    importlib_metadata~=1.7 # We could work with 3.1, but argparse needs <2
-    importlib_resources~=1.4
+    importlib_metadata>=1.5,<4;python_version<="3.7"
```
Do we not need to up the min requirements for argcomplete too?

mjpieters replied:
Not really. `argcomplete~=1.10` is the same thing as `argcomplete>=1.10,<2`, and so `1.22` is included in that range.

mjpieters replied:
Put differently: that'd be a separate task if someone wanted to update this explicitly. But unless you depend on the bugfixes in later releases, I'd leave this be. No need to make the dependency link tighter.

ashb replied:
Historical we've updated the intermediary rep, otherwise I thought:

Pip would leave argparse at 1.10, which has a dep of importlib_metadata, but then this new dep would install a newer importlib_metadata, leading to incompatible version errors.

Particularly for "older" pips. Am I misremembering?

mjpieters replied:
When you use `pip install -U Apache-airflow` to upgrade airflow, a newer `argcomplete` would be installed, automatically.

In my experience it takes installing additional packages with incompatible dependencies for things to break; newer `pip` versions warn when this happens however. Running `pip check` lets you list the problem packages.

potiuk replied:
The problem might be that in our process (in master merges) we automatically do 'pip install .[all] --upgrade --upgrade-strategy eager` to automatically find set of latest constraints that are matching setup.py + setup.cfg limitations, Then we run pip check and then we run all tests and if it all passes we automatically commit the new constraints.

As far as I remember without ~1.* for importlib_resources in install_requires caused a problem as pip first tries to install the highest version matching in `install_requires` only - without looking at extras. Once it installs this version, it will only then try to do eager update on all other extras and one of our dependencies already has importib_resources ~= 1.4. Eager upgrade will never downgrade the packages installed by install_requires, so this results in version conflict.

mjpieters replied:
Thinking about this some more: we did loosen the restrictions on what versions are acceptable but `importlib_metadata` *is* a new dependency relative to 2.0.0b3. If you then use `pip install apache-airflow>2.0.0b3` (say) *without* `-U` then I think you could indeed end up with an importlib_metadata that is incompatible with an existing argcomplete version if you only have argparse 1.10.

Want me to up the version on that package too, then?

potiuk replied:
In order to test it out, I've added the labels "upgrade to latest dependencies" and "full tests needed". Can you please @mjpieters amend and push --force-with-lease to test my theory? 

With those labels, the PR will automatically try to do what master merge does - it will try to figure out latest constraints with eager upgrade strategy and it will do it for all python versions, so let's see if it works.

potiuk replied:
BTW. We do not recommend running unconstrained upgrade of airflow - this has caused us terrible problems in the past. That's why we have this constraint mechanism and with the recent changes (I am also fixing that for 1.10.14) we take care about making sure that the set of constraints that we recommend (in INSTALL.txt) are non-conflicting and tested (and we update the constraints-1-10 branch and constrains-master automatically to include latest releases of dependencies (all tests muss pass and pip check must show no-conflicts for those as of Friday)

potiuk replied:
Confirmed: @mjpieters @ashb 

importlib_resources~=1.4 is a requirement of cnf-lint which is requirement of moto 1.3.16 which we have. So anything above 1 will cause version conflict. mot 1.3.16 is the latest available version, so I do not think we can do anything about it.

```
moto==1.3.16
  - aws-xray-sdk [required: >=0.93,!=0.96, installed: 2.6.0]
    - botocore [required: >=1.11.3, installed: 1.18.18]
      - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
      - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
        - six [required: >=1.5, installed: 1.15.0]
      - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
    - future [required: Any, installed: 0.18.2]
    - jsonpickle [required: Any, installed: 1.4.1]
      - importlib-metadata [required: Any, installed: 1.7.0]
        - zipp [required: >=0.5, installed: 3.4.0]
    - wrapt [required: Any, installed: 1.12.1]
  - boto [required: >=2.36.0, installed: 2.49.0]
  - boto3 [required: >=1.9.201, installed: 1.15.18]
    - botocore [required: >=1.18.18,<1.19.0, installed: 1.18.18]
      - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
      - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
        - six [required: >=1.5, installed: 1.15.0]
      - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
    - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
    - s3transfer [required: >=0.3.0,<0.4.0, installed: 0.3.3]
      - botocore [required: >=1.12.36,<2.0a.0, installed: 1.18.18]
        - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
        - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
          - six [required: >=1.5, installed: 1.15.0]
        - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
  - botocore [required: >=1.12.201, installed: 1.18.18]
    - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
    - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
      - six [required: >=1.5, installed: 1.15.0]
    - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
  - cfn-lint [required: >=0.4.0, installed: 0.42.0]
    - aws-sam-translator [required: >=1.25.0, installed: 1.31.0]
      - boto3 [required: ~=1.5, installed: 1.15.18]
        - botocore [required: >=1.18.18,<1.19.0, installed: 1.18.18]
          - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
          - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
            - six [required: >=1.5, installed: 1.15.0]
          - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
        - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
        - s3transfer [required: >=0.3.0,<0.4.0, installed: 0.3.3]
          - botocore [required: >=1.12.36,<2.0a.0, installed: 1.18.18]
            - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
            - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
              - six [required: >=1.5, installed: 1.15.0]
            - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
      - jsonschema [required: ~=3.2, installed: 3.2.0]
        - attrs [required: >=17.4.0, installed: 20.3.0]
        - importlib-metadata [required: Any, installed: 1.7.0]
          - zipp [required: >=0.5, installed: 3.4.0]
        - pyrsistent [required: >=0.14.0, installed: 0.17.3]
        - setuptools [required: Any, installed: 50.3.2]
        - six [required: >=1.11.0, installed: 1.15.0]
      - six [required: ~=1.15, installed: 1.15.0]
    - importlib-resources [required: ~=1.4, installed: 1.5.0]
```

ashb replied:
Yet again pip's lack of a fully featured dependency solver bites us

mjpieters replied:
> In order to test it out, I've added the labels "upgrade to latest dependencies" and "full tests needed". Can you please @mjpieters amend and push --force-with-lease to test my theory?

I rebased and pushed before I noticed that a bot had removed the tags again.

I've pushed a change now that ensures we only allow `importlib-resources` 1.x (starting with `1.4`)

potiuk replied:
yeah - full tests needed only works after the approval status is "Green" (whch is actually good - we can see if it wiorks for 3.6 first and once it does, we approve and the label will be added and we rerun it again with full matrix of tests.

potiuk replied:
Let's see if it works. I will approve it tentatively, and please rebase @mjpieters - let's see if it will work.

ashb reviewed (commented):
I think there's a use of importlib_resources that should be correct too.

mjpieters committed:
Narrow importlib_resources back down to a 1.x release

There are provider dependencies that can't use anything newer, and pip is not sophisticated enough to downgrade packages based on new information.

potiuk reviewed (approved):
Approved to get the label working. But let's not merge it until IP check succeeds for all python versions.
